### PR TITLE
Marketplace: update no results page

### PR DESCRIPTION
**Pull request by raicem**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:
We show two separate no results components on the search results page, this PR reduces it to one. 

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Checkout this branch and build changes
2. Go to any In App Marketplace page (Discover, Browse, Themes... doesn't matter). 
3. Perform a search where you'll get hits for both extensions and themes. 
    - You should see both results for extensions and themes on the search reuslts page. 
    - Clicking **View all** should show you the full result set. 
4. Perform a search where you'll get hits only for extensions.
    -  You'll only see the extensions component. There will no mention of themes.
    - The list will be expanded: there won't be any **View all** button and you'll see the category selector.
5. Perform a search where you'll get hits only for themes.
    -  You'll only see the themes component. There will no mention of extensions.
    - The list will be expanded: there won't be any **View all** button and you'll see the category selector.
6. Perform a search where you'll get **no** hits for extensions and themes.
    -   You'll see the new NoResults component.
    - You should see recommendations for both themes and extensions.
7. Browse **Browse** and **Themes** pages. Confirm you can use filters still. 
8. Please test around these changes.

### Screenshots
image

### CODE REVIEW

**Comment by raicem** [comment]
I found this bug. Go to **Browse** page, search for **Watermark** and then switch the category to **Payments**. That'll leave you with no extension results. But our no results page will still say "No extensions or themes found…" and recommend themes. Not ideal. We were not looking for themes.

**Review by andfinally — COMMENTED** [review]
Ah this is fantastic work, thanks for rustling this up so quickly! Going to give it a closer look, but when testing, I didn't get the bug you saw – I got WooPayments when trying to reproduce it.

The only issue I've found so far is that the CSS we were using to hide the fourth item in the suggestions on mid-sized viewports doesn't seem to be working any more.

image

**Comment by andfinally** [comment]
I think we can fix the item count in the "no results" suggestions with this.

```diff
Index: plugins/woocommerce-admin/client/marketplace/components/product-list-content/product-list-content.scss
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/plugins/woocommerce-admin/client/marketplace/components/product-list-content/product-list-content.scss b/plugins/woocommerce-admin/client/marketplace/components/product-list-content/product-list-content.scss
--- a/plugins/woocommerce-admin/client/marketplace/components/product-list-content/product-list-content.scss	(revision 7285e8c1c9b30c43728920d169d554ef7cf75aac)
+++ b/plugins/woocommerce-admin/client/marketplace/components/product-list-content/product-list-content.scss	(date 1700582099827)
@@ -21,9 +21,11 @@
 			grid-template-columns: repeat(2, 1fr);
 		}
 
-		// Hide third and above product cards on Discover and search results page due to API result count
-		// These are progressively displayed at larger screen sizes.
-		&__discover .woocommerce-marketplace__product-card:nth-child(n+3) {
+		// Hide third and above product cards on Discover and suggestions on "no results" search results page
+		// due to item count. We display progressively more on larger screen sizes.
+		&__discover .woocommerce-marketplace__product-card:nth-child(n+3),
+		&__no-results .woocommerce-marketplace__product-card:nth-child(n+3)
+		{
 			display: none;
 		}
 	}
@@ -35,7 +37,8 @@
 			gap: $large-gap;
 			grid-template-columns: repeat(3, 1fr);
 		}
-		&__discover .woocommerce-marketplace__product-card:nth-child(3) {
+		&__discover .woocommerce-marketplace__product-card:nth-child(3),
+		&__no-results .woocommerce-marketplace__product-card:nth-child(3) {
 			display: block;
 		}
 	}
@@ -46,7 +49,8 @@
 		&__product-list-content {
 			grid-template-columns: repeat(4, 1fr);
 		}
-		&__discover .woocommerce-marketplace__product-card:nth-child(4) {
+		&__discover .woocommerce-marketplace__product-card:nth-child(4),
+		&__no-results .woocommerce-marketplace__product-card:nth-child(4) {
 			display: block;
 		}
 	}
```

File: plugins/woocommerce-admin/client/marketplace/components/search-results/search-results.tsx [hunk 1]
```diff
@@ -2,13 +2,21 @@
...
  */
 import './search-results.scss';
 import { Product, ProductType, SearchResultType } from '../product-list/types';
 import Products from '../products/products';
+import NoResults from '../product-list-content/no-results';
+import { MarketplaceContext } from '../../contexts/marketplace-context';
+import {
+	MARKETPLACE_ITEMS_PER_PAGE,
+	MARKETPLACE_SEARCH_RESULTS_PER_PAGE,
+} from '~/marketplace/components/constants';
```
Comment (andfinally) [inline plugins/woocommerce-admin/client/marketplace/components/search-results/search-results.tsx]: Bero's VSCode was generating these dodgy paths, I think – maybe the same thing is happening with yours.

```suggestion
import {
	MARKETPLACE_ITEMS_PER_PAGE,
	MARKETPLACE_SEARCH_RESULTS_PER_PAGE,
} from '../../components/constants';
```

File: plugins/woocommerce-admin/client/marketplace/components/search-results/search-results.tsx [hunk 2]
```diff
@@ -22,50 +30,116 @@ export default function SearchResults( props: SearchResultProps ): JSX.Element {
...
-		/>
-	);
+	function productsComponent(
+		products: Product[],
+		type: ProductType,
+		overrides: {
+			categorySelector?: boolean;
+			showAllButton?: boolean;
+			perPage?: number;
+		} = {}
```
Comment (andfinally) [inline plugins/woocommerce-admin/client/marketplace/components/search-results/search-results.tsx]:
```suggestion
		overrides: Overrides = {}
```

File: plugins/woocommerce-admin/client/marketplace/components/product-list-content/no-results.tsx [hunk 3]
```diff
@@ -1,159 +1,152 @@
...
+	const [ isLoading, setIsLoading ] = useState( false );
 	const query = useQuery();
 	const showCategorySelector = query.tab === 'search' && query.section;
+	const productGroupsForSearchType: {
+		[ key in SearchResultType ]: string[];
+	} = {
+		[ SearchResultType.all ]: [ 'most-popular', 'popular-themes' ],
+		[ SearchResultType.theme ]: [ 'popular-themes' ],
+		[ SearchResultType.extension ]: [ 'most-popular' ],
+	};
```
Comment (andfinally) [inline plugins/woocommerce-admin/client/marketplace/components/product-list-content/no-results.tsx]: TypeScript linter in my IDE complains about the type here:

```
Initializer type {} is not assignable to variable type {[key in SearchResultType]: string[]} 
```

It suggests two alternatives which make it happy. The first is:

```ts
	const productGroupsForSearchType: {} = {
		[ SearchResultType.all ]: [ 'most-popular', 'popular-themes' ],
		[ SearchResultType.theme ]: [ 'popular-themes' ],
		[ SearchResultType.extension ]: [ 'most-popular' ]
	};
```

The second is this:

```ts
	const productGroupsForSearchType: {
		[ key in SearchResultType ]: string[];
	} = {
		all: [], extension: [], theme: [],
		[ SearchResultType.all ]: [ 'most-popular', 'popular-themes' ],
		[ SearchResultType.theme ]: [ 'popular-themes' ],
		[ SearchResultType.extension ]: [ 'most-popular' ]
	};
```

I'm not sure why it doesn't like your type definition – maybe it'll make more sense to you. I guess the first solution makes the type less well defined.

**Review by andfinally — COMMENTED** [review]
We're not using the `MARKETPLACE_SEARCH_RESULTS_PER_PAGE` imported in `products.tsx`.

I managed to reproduce the bug by searching for "deco", seeing one theme result, and clicking on the `Fashion and Apparel` category. Maybe the solution is to still show `MARKETPLACE_SEARCH_RESULTS_PER_PAGE` of results with the "View all" button. But I don't think this is a fatal bug, personally. I reckon we could maybe go with it and come back and fix it later.

File: plugins/woocommerce-admin/client/marketplace/components/product-list-content/no-results.tsx (hunk 3 above)
Comment (raicem) [reply plugins/woocommerce-admin/client/marketplace/components/product-list-content/no-results.tsx]: That is odd. I'm not getting anything on VSCode or on PHPStorm. I'll remove the explicit type definition. It won't change how the code is being run. 

```suggestion
	const productGroupsForSearchType = {
		[ SearchResultType.all ]: [ 'most-popular', 'popular-themes' ],
		[ SearchResultType.theme ]: [ 'popular-themes' ],
		[ SearchResultType.extension ]: [ 'most-popular' ],
	};
```

File: plugins/woocommerce-admin/client/marketplace/components/search-results/search-results.tsx (hunk 1 above)
Comment (raicem) [reply plugins/woocommerce-admin/client/marketplace/components/search-results/search-results.tsx]: Ahh, thanks! I won't accept it over GitHub but will do a commit with you as the co-author.

**Review by raicem — COMMENTED** [review]
Thanks for the review @andfinally!

1) Addressed your 3 suggetions: https://github.com/woocommerce/woocommerce/pull/41603/commits/01a34c590c65ada639c21169a504f3b3bba6f978
2) Removed unused `MARKETPLACE_SEARCH_RESULTS_PER_PAGE` in `products.tsx`
3) Applied your patch for the card item count in no results page. I tried this as well but caused bugs. But your version worked perfect. So it's up at https://github.com/woocommerce/woocommerce/pull/41603/commits/22f470b54dc4762cd8e68fd8a9d7b705fb4062dc

File: plugins/woocommerce-admin/client/marketplace/components/search-results/search-results.tsx [hunk 4]
```diff
@@ -22,50 +30,112 @@ export default function SearchResults( props: SearchResultProps ): JSX.Element {
...
-			return extensionComponent;
+			return extensionsComponent( { showAllButton: false } );
 		}
 
-		if ( extensionList.length === 0 && themeList.length > 0 ) {
+		if ( query?.section === SearchResultType.theme ) {
+			return themesComponent( { showAllButton: false } );
+		}
+
+		// Components can handle their isLoading state. So we can put them both on the page.
```
Comment (kdevnel) [inline plugins/woocommerce-admin/client/marketplace/components/search-results/search-results.tsx:86]: These comments are a really great example of useful context. I could see it being easy to miss that the loading state is here, but that the components handle how they appear when loading separately.

**Review by kdevnel — APPROVED** [review]
LGTM!

File: plugins/woocommerce-admin/client/marketplace/components/search-results/search-results.tsx (hunk 4 above)
Comment (raicem) [reply plugins/woocommerce-admin/client/marketplace/components/search-results/search-results.tsx:86]: Thanks! This is one of the parts of the code I'm not happy with, for the reason you have pointed out. It's very easy to miss. Comments for the rescue until we fix this.

**Review by andfinally — APPROVED** [review]
Looks great, thanks for the tweaks! 💯

**Pull request merged**